src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -0,0 +1,305 @@
+#include <stdexcept>
+
+#include <boost/foreach.hpp>
+
+#include "zcash/IncrementalMerkleTree.hpp"
+#include "crypto/sha256.h"
+#include "zerocash/utils/util.h" // TODO: remove these utilities
+
+namespace libzcash {
+
+SHA256Compress SHA256Compress::combine(const SHA256Compress& a, const SHA256Compress& b)
+{
+    // This is a performance optimization.
+    if (a.IsNull() && b.IsNull()) {
+        return a;
+    }
+
+    SHA256Compress res = SHA256Compress();
+
+    CSHA256 hasher;
+    hasher.Write(a.begin(), 32);
+    hasher.Write(b.begin(), 32);
+    hasher.FinalizeNoPadding(res.begin());
+
+    return res;
+}
+
+
+template <typename Hash>
+class PathFiller {
+private:
+    std::deque<Hash> queue;
+public:
+    PathFiller() : queue() { }
+    PathFiller(std::deque<Hash> queue) : queue(queue) { }
+
+    Hash next() {
+        if (queue.size() > 0) {
+            Hash h = queue.front();
+            queue.pop_front();
+
+            return h;
+        } else {
+            return Hash();
+        }
+    }
+};
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::wfcheck() const {
+    if (parents.size() >= Depth) {
+        throw std::ios_base::failure("tree has too many parents");
+    }
+
+    // The last parent cannot be null.
+    bool wasnull = false;
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
+        wasnull = !parent;
+    }
+
+    if (wasnull) {
+        throw std::ios_base::failure("tree has non-canonical representation of parent");
+    }
+
+    // Left cannot be empty when right exists.
+    if (!left && right) {
+        throw std::ios_base::failure("tree has non-canonical representation of tree");
+    }
+
+    // Left cannot be empty when parents is nonempty.
+    if (!left && parents.size() > 0) {
+        throw std::ios_base::failure("tree has non-canonical representation of tree");
+    }
+}
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::append(Hash obj) {
+    if (is_complete(Depth)) {
+        throw std::runtime_error("tree is full");
+    }
+
+    if (!left) {
+        // Set the left leaf
+        left = obj;
+    } else if (!right) {
+        // Set the right leaf
+        right = obj;
+    } else {
+        // Combine the leaves and propagate it up the tree
+        boost::optional<Hash> combined = Hash::combine(*left, *right);
+
+        // Set the left leaf to the object and make the right object none
+        left = obj;
+        right = boost::none;
+
+        for (size_t i = 0; i < Depth; i++) {
+            if (i < parents.size()) {
+                if (parents[i]) {
+                    combined = Hash::combine(*parents[i], *combined);
+                    parents[i] = boost::none;
```

Comment:
I think I also understand why we want to keep `right` (referring the @defuse's comment in the discussion linked above): it makes `is_complete` easier to write because the final fully-filled tree state consists of `left`, `right` and `parents` being completely full. Without `right`, the fully-filled tree state would be `left` empty and `parents` empty except for the last element, which would not be as trivial to state.